Be nice to Circle (enforce Gradle, React-Native Bundler parallelism limits)

hawkrives opened the pull request:
Hopefully this will stop the spurious Android build errors we've been seeing on Circle.

I say merge this, if the build passes. If this slows down our local dev too much, we can play with increasing it a bit.

- - -

Additional stuff by @rye:

- Further reduced gradle limits to `512m`.  This _should_, ya know, be enough.  But we all know how that goes.

- Noticed when running the failing builds over SSH that a bunch of `node` processes were spinning up when the bundler was running, which is why that task always fails.  I hypothesize that this is because we don't _actually_ have limits on CircleCI when the container runs, so Node sees _lots_ of CPUs. (like, 64 or 32 of them)  [Metro](https://github.com/facebook/metro/blob/0f7ad193c75183eeff1b356644ccf22b0813bb04/packages/metro/src/commands/build.js#L39) uses some stupid syntax to check what the CPU count is, and so happily kicks off 64 or whatever workers. (which probably frightens Circle which gets stabby and kills all of our workers, thus ending gradle with no notice)

  - This will reduce the overall possible parallelism but will also ensure that we keep well under our limit.  Currently the new maximum if the `CI` variable is set is `2` workers.

hawkrives committed:
drop Gradle's allotted memory size

hawkrives committed:
add -XX:+HeapDumpOnOutOfMemoryError to gradle

hawkrives committed:
check version of node on android builds

rye committed:
Merge branch 'master' into lower-gradle-ram

rye committed:
Reduce Gradle RAM allowance even further

Stop being so hungry, Gradle.  (Also, make way for React-Native, apparently)

rye committed:
Pass `--max-workers` to the react-native bundler

Currently this does not have any check to only do this if CI is
set---I'm working on that right now.  This is just a check to see if
this works.

rye committed:
Merge branch 'lower-gradle-ram' of https://github.com/StoDevX/AAO-React-Native into lower-gradle-ram

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives commented on android/app/react.gradle:
```diff
@@ -68,7 +68,7 @@ task bundleReleaseJsAndAssets(type: Exec) {
     } else {
         commandLine "react-native", "bundle", "--platform", "android", "--dev", "false", "--entry-file",
-            entryFile, "--bundle-output", jsBundleFileRelease, "--assets-dest", resourcesDirRelease
+            entryFile, "--max-workers", "2", "--bundle-output", jsBundleFileRelease, "--assets-dest", resourcesDirRelease
```
I suppose there's no real need for us to run at full parallelism locally, either.

rye committed:
Attempt to use a ternary switch on the bundler commandline

hawkrives reviewed (commented):
Approooooove (I started the pr so I can't)

rye reviewed (dismissed):
I approve, bwahahaha 😈

hawkrives commented on android/app/react.gradle:
```diff
@@ -68,7 +68,7 @@ task bundleReleaseJsAndAssets(type: Exec) {
     } else {
         commandLine "react-native", "bundle", "--platform", "android", "--dev", "false", "--entry-file",
-            entryFile, "--max-workers", "2", "--bundle-output", jsBundleFileRelease, "--assets-dest", resourcesDirRelease
+            entryFile, "--bundle-output", jsBundleFileRelease, "--assets-dest", resourcesDirRelease, (System.getenv("CI") == "true") ? "-j2" : ""
```
Nitpick: `System.getenv("CI") ?: "" != ""`? To handle systems that don't set CI to true and to make sure that gradle doesn't throw when not defined?

rye replied:
Is that what the condition should be or the entire ternary?

hawkrives replied:
The entire ternary. ?: is the "Elvis" operator; if the lhs doesn't exist / is falsy, it uses the right side. 

Bah. That's not really what I want. I want System.hasEnv, but I don't think that exists.

rye committed:
Use better condition in ternary

rye reviewed (dismissed):
I approve (again)

rye committed:
Fix the ternary to make it actually valid

rye committed:
Add the org.gradle.workers.max configuration variable

This also makes sure that Gradle doesn't decide to... spawn 32 workers
for its nefarious activities.

rye committed:
Bump up the maximum RAM for Gradle again

Since we've narrowed the issue down to the RN bundler, this can happen now.

hawkrives reviewed (commented):
🚢

rye reviewed (approved):
Very good.  We can always revisit parallelism issues if they pop up later.
